pulsar/internal/connection.go

```diff
@@ -456,10 +474,19 @@ func (c *connection) runPingCheck(pingCheckTicker *time.Ticker) {
 	}
 }
 
-func (c *connection) WriteData(ctx context.Context, data Buffer) {
+func (c *connection) WriteData(ctx context.Context, data *SharedBuffer) {
+	written := false
+	defer func() {
+		if !written {
+			// Buffer has failed to be written and can now be reused.
+			c.bufferPool.Put(data)
```

Comment:
Can we add a new method like `Recycle` into the `SharedBuffer` so we don't need to maintain the bufferPool in the `connection`. The SharedBuffer can hold the ref to the buffer pool.